agent/exec/containerd/adapter.go

```diff
@@ -0,0 +1,682 @@
+// are mostly naked calls to the client API, seeded with information from
+// containerConfig.
+type containerAdapter struct {
+	conn              *grpc.ClientConn
+	client            execution.ContainerServiceClient
+	container         *api.ContainerSpec
+	task              *api.Task
+	secrets           exec.SecretGetter
+	dir               string
+	resolvedImageName string
+	deleteResponse    *execution.DeleteResponse
+}
+
+func newContainerAdapter(conn *grpc.ClientConn, containerDir string, task *api.Task, secrets exec.SecretGetter) (*containerAdapter, error) {
+	container := task.Spec.GetContainer()
+	if container == nil {
+		return nil, exec.ErrRuntimeUnsupported
+	}
+
+	dir := filepath.Join(containerDir, task.ID)
+
+	return &containerAdapter{
+		conn:      conn,
+		client:    execution.NewContainerServiceClient(conn),
+		container: container,
+		task:      task,
+		secrets:   secrets,
+		dir:       dir,
+	}, nil
+}
+
+func (c *containerAdapter) applyLayer(ctx context.Context, cs content.Store, rootfs string, layer digest.Digest) error {
+	blob, err := cs.Reader(ctx, layer)
+	if err != nil {
+		return err
+	}
+
+	rd, err := compression.DecompressStream(blob)
+	if err != nil {
+		return err
+	}
+
+	_, err = archive.Apply(ctx, rootfs, rd)
+
+	blob.Close()
+	return err
+}
+
+// github.com/containerd/containerd cmd/ctr/utils.go, dropped stdin handling
+func prepareStdio(stdout, stderr string, console bool) (*sync.WaitGroup, error) {
+	var wg sync.WaitGroup
+	ctx := context.Background()
+
+	f, err := fifo.OpenFifo(ctx, stdout, syscall.O_RDONLY|syscall.O_CREAT|syscall.O_NONBLOCK, 0700)
+	if err != nil {
+		return nil, err
+	}
+	defer func(c io.Closer) {
+		if err != nil {
+			c.Close()
+		}
+	}(f)
+	wg.Add(1)
+	go func(r io.ReadCloser) {
+		io.Copy(os.Stdout, r)
+		r.Close()
+		wg.Done()
+	}(f)
+
+	f, err = fifo.OpenFifo(ctx, stderr, syscall.O_RDONLY|syscall.O_CREAT|syscall.O_NONBLOCK, 0700)
+	if err != nil {
+		return nil, err
+	}
+	defer func(c io.Closer) {
+		if err != nil {
+			c.Close()
+		}
+	}(f)
+	if !console {
+		wg.Add(1)
+		go func(r io.ReadCloser) {
+			io.Copy(os.Stderr, r)
+			r.Close()
+			wg.Done()
+		}(f)
+	}
+
+	return &wg, nil
+}
+
+func (c *containerAdapter) pullImage(ctx context.Context) error {
+	options := docker.ResolverOptions{}
+
+	tr := &http.Transport{
+		Proxy: http.ProxyFromEnvironment,
+		DialContext: (&net.Dialer{
+			Timeout:   30 * time.Second,
+			KeepAlive: 30 * time.Second,
+			DualStack: true,
+		}).DialContext,
+		MaxIdleConns:          10,
+		IdleConnTimeout:       30 * time.Second,
+		TLSHandshakeTimeout:   10 * time.Second,
+		TLSClientConfig:       &tls.Config{},
+		ExpectContinueTimeout: 5 * time.Second,
+	}
+
+	options.Client = &http.Client{
+		Transport: tr,
+	}
+
+	resolver := docker.NewResolver(options)
+
+	name, desc, fetcher, err := resolver.Resolve(ctx, c.container.Image)
+	if err != nil {
+		return errors.Wrap(err, "Failed to resolve ref")
```

Comment:
The preferred style in the Go world is for for errors to start with lowercase characters, but we don't follow this consistently in the Docker ecosystem.